// match-index/lib/match-index.js
// LICENSE : MIT
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var flagsGetter = require("regexp.prototype.flags");
var assert = require("assert");
/**
 * @param {string} text
 * @param {RegExp} regExp regExp should include capture
 * @returns {MatchCaptureGroup[]} return array of MatchCaptureGroup
 */
function matchCaptureGroupAll(text, regExp) {
    var source = regExp.source;
    assert(source.indexOf("(") >= 0, "RegExp should contains capture group at least one");
    var all = matchAll(text, regExp);
    var captureGroups = [];
    all.forEach(function (match) {
        match.captureGroups.forEach(function (captureGroup) {
            captureGroups.push(captureGroup);
        });
    });
    return captureGroups;
}
exports.matchCaptureGroupAll = matchCaptureGroupAll;
/**
 * matchAll function inspired String.prototype.matchAll
 * @param {String} text
 * @param {RegExp} regExp
 * @returns {MatchAllGroup[]}
 * @see reference https://github.com/tc39/String.prototype.matchAll
 * http://stackoverflow.com/questions/15934353/get-index-of-each-capture-in-a-javascript-regex
 */
function matchAll(text, regExp) {
    var matches = [];
    var flags = regExp.flags || flagsGetter(regExp);
    if (flags.indexOf('g') === -1) {
        flags = 'g' + flags;
    }
    var rx = new RegExp(regExp.source, flags);
    text.replace(rx, function () {
        var matchAll = Array.prototype.slice.call(arguments, 0, -2);
        var match = {
            all: matchAll,
            input: arguments[arguments.length - 1],
            index: arguments[arguments.length - 2],
            captureGroups: []
        };
        var groups = matchAll.slice(1);
        var captureGroups = [];
        for (var cursor = match.index, l = groups.length, i = 0; i < l; i++) {
            var index = cursor;
            if (i + 1 !== l && groups[i] !== groups[i + 1]) {
                var nextIndex = text.indexOf(groups[i + 1], cursor);
                while (true) {
                    var currentIndex = text.indexOf(groups[i], index);
                    if (currentIndex !== -1 && currentIndex <= nextIndex) {
                        index = currentIndex + 1;
                    }
                    else {
                        break;
                    }
                }
                index--;
            }
            else {
                index = text.indexOf(groups[i], cursor);
            }
            cursor = index + (groups[i] ? groups[i].length : 0);
            var captureGroup = {
                text: groups[i],
                index: index
            };
            captureGroups.push(captureGroup);
        }
        match.captureGroups = captureGroups;
        matches.push(match);
        /*
            index,
            input,
            all,
            captureGroups = [{
                text,
                index
            }]
         */
        return "";
    });
    return matches;
}
exports.matchAll = matchAll;
